### ISSUE TITLE
Should there be an intermediate interface

### ISSUE THREAD

**Issue by abelsiqueira** [abelsiqueira/CUTEst.jl#4]
`CUTEst.ufn(st, n, x, f)` is the raw call, that requires `st` and `n` to be `Ptr{Int32}` and `x` and `f` to be `Ptr{Float64}`.
`f = CUTEsf.obj_fun(prob, x)` is (will be) the high level call, that requires `x` to be an array of correct size that will be converted to the specific type, then f will be returned as that same type. In this type of function, copies and conversions will be made.

Is there an use for a function
`CUTEst.jl_ufn(st, n, x, f)` where there is no type requirement to the arguments and copies and conversions would be made?

**Comment by dpo** [comment]
It is useful t have the expert call handy, but to a Julia user, I feel like a call of the form
````
obj(nlp::CUTEstProb, x::Array{Float64,1})
````
would be more useful. We should feel free to modernize the CUTEst names in a convenience interface.

There's no need to pass `n`; it's implicitly in `x` (and in `nlp`). There's also no need for `st`; if it's nonzero, an error should be raised.

I did something like that in [ampl.jl](https://github.com/dpo/ampl.jl). See, e.g., [this line](https://github.com/dpo/ampl.jl/blob/master/src/ampl.jl#L228).

Perhaps we could also make `CUTEstProb` more uniform with `AmplModel`?

**Comment by abelsiqueira** [comment]
Hello @dpo, thanks for the comment. I do intend to make a Julia-friendly interface in the near future, I was just wondering about a not-so-friendly-not-so-expert function **in addition** to those.

About uniformity with `AmplModel`, I will take a closer look at it some time, but it looks like at least a conversion function is very doable.

**Comment by dpo** [comment]
@abelsiqueira In the past few days, I've been cleaning up my [ampl.jl](https://github.com/dpo/ampl.jl) and I extracted out of it a type to represent a generic optimization problem, which is now the base of [NLP.jl](https://github.com/optimizers/NLP.jl). I've had success with that sort of model in the past in other languages. I remodeled my earlier gist to use NLP.jl and put it in [CUTEst.jl](https://github.com/optimizers/CUTEst.jl).

Advantages of my approach are:
- a consistent Julia type across AMPL and CUTEst models
- an interface that looks like the Matlab CUTEst interface, which people are used to (though I would also like to have an interface that looks like the one in ampl.jl)
- each problem has its own shared library. For now, I still can't have multiple problems decoded at once, but that's what I'd like to do down the road.

Now I don't have your expert interface, which I think would be useful to expert users, and I've only interfaced a few functions.

Should we join efforts?

**Comment by abelsiqueira** [comment]
It also looks like you make a better job on decoding.

Joining seems like a good option. What do you suggest?

**Comment by dpo** [comment]
The interface I wrote is strongly inspired by yours! I'm proposing to give you write access to my `CUTEst.jl` and we can work with pull requests (they make it easy to comment on code before we merge). Among the things we need:

- your expert interface
- documentation (I suggest Docile.jl and Lexicon.jl; I'm using them in other projects and they're great)
- a more complete convenience interface.

Does that work?

**Comment by abelsiqueira** [comment]
You're proposing to continue the work on yours. I was also gonna suggest that, since it's easier to port the expert interface.

But I was trying to test your package and received a segmentation fault. Do you think you can put up a Travis to test it? If not, maybe changing my repository will be less work.

About your suggestions, we can work with that.

**Comment by dpo** [comment]
I'll definitely set up Travis builds, but they're so difficult to debug that I'd rather try by hand first. I supposed you built your shared CUTEst library using your package? It would be great if we could make this work on both OSX and Linux.

**Comment by abelsiqueira** [comment]
Yes, I used my package. Basically, I compiled almost every tool in $CUTEST/src/tools/ with -fPIC and link with -shared to my /usr/lib. It works with a C test and with my package, but I haven't tested in OSX. I'll try to make my package work on OSX on Travis.

**Comment by dpo** [comment]
The Homebrew formula already builds a shared library for OSX. I'd like to make it work on Linux too. I notice that you use `ld` to build your shared lib, while I use `gfortran -shared`.

**Comment by abelsiqueira** [comment]
I use ld -shared [...] -lgfortran. I tested using gfortran and it also works, so this shouldn't be an issue.

**Comment by abelsiqueira** [comment]
To build on travis with osx I need to make a requirement by e-mail. I'm waiting on the reply for the moment. In the mean time, my package works on your OSX?
To make it work, you need a `libcutest.so` "visible", change to branch `osx` and issue the commands
````
julia> Pkg.rm("CUTEst")
julia> Pkg.clone(pwd())
julia> Pkg.build("CUTEst")
julia> Pkg.test("CUTEst")
````
Thanks

**Comment by abelsiqueira** [comment]
Renamed my/this repository to old-CUTEst.jl, and will fork yours.
